### PR TITLE
SONAR-8995 Scanner fails when using issue exclusions

### PR DESCRIPTION
**Pull request by dbmeneses**

### CODE REVIEW

File: sonar-scanner-engine/src/main/java/org/sonar/scanner/issue/ignore/scanner/IssueExclusionsRegexpScanner.java [hunk 1]
```diff
@@ -76,30 +78,32 @@ private void init() {
...
-          exclusionPatternInitializer.getPatternMatcher().addPatternToExcludeResource(resource);
-          // nothing more to do on this file
-          LOG.debug("- Exclusion pattern '{}': every violation in this file will be ignored.", pattern);
-          return;
+        // first check the single regexp patterns that can be used to totally exclude a file
+        for (java.util.regex.Pattern pattern : allFilePatterns) {
+          if (pattern.matcher(line).find()) {
+            exclusionPatternInitializer.getPatternMatcher().addPatternToExcludeResource(resource);
+            // nothing more to do on this file
+            LOG.debug("- Exclusion pattern '{}': every violation in this file will be ignored.", pattern);
```
Comment (henryju) [inline sonar-scanner-engine/src/main/java/org/sonar/scanner/issue/ignore/scanner/IssueExclusionsRegexpScanner.java]: s/violation/issues

File: sonar-scanner-engine/src/main/java/org/sonar/scanner/issue/ignore/scanner/IssueExclusionsRegexpScanner.java [hunk 2]
```diff
@@ -76,30 +78,32 @@ private void init() {
     currentLineExclusion = null;
   }
 
-  public void scan(String resource, File file, Charset sourcesEncoding) throws IOException {
+  public void scan(String resource, Path filePath, Charset encoding) throws IOException {
```
Comment (janos-ss) [inline sonar-scanner-engine/src/main/java/org/sonar/scanner/issue/ignore/scanner/IssueExclusionsRegexpScanner.java]: I don't see where `encoding` is used. Previously it was used in `Files.readAllLines`. Is that ok?

File: sonar-scanner-engine/src/main/java/org/sonar/scanner/issue/ignore/scanner/IssueExclusionsRegexpScanner.java [hunk 3]
```diff
@@ -76,30 +78,32 @@ private void init() {
...
     init();
 
-    List<String> lines = Files.readAllLines(file.toPath(), sourcesEncoding);
     int lineIndex = 0;
-    for (String line : lines) {
-      lineIndex++;
-      if (line.trim().length() == 0) {
-        continue;
-      }
+    try (BufferedReader br = Files.newBufferedReader(filePath)) {
```
Comment (henryju) [inline sonar-scanner-engine/src/main/java/org/sonar/scanner/issue/ignore/scanner/IssueExclusionsRegexpScanner.java]: Use charset!!

File: sonar-scanner-engine/src/main/java/org/sonar/scanner/issue/ignore/scanner/IssueExclusionsLoader.java [hunk 4]
```diff
@@ -47,27 +45,29 @@ public boolean shouldExecute() {
...
+  public void preLoadAllFiles() {
     for (InputFile inputFile : fileSystem.inputFiles(fileSystem.predicates().all())) {
-      try {
-        String componentEffectiveKey = ((DefaultInputFile) inputFile).key();
-        if (componentEffectiveKey != null) {
-          String path = inputFile.relativePath();
-          inclusionPatternInitializer.initializePatternsForPath(path, componentEffectiveKey);
-          exclusionPatternInitializer.initializePatternsForPath(path, componentEffectiveKey);
-          if (exclusionPatternInitializer.hasFileContentPattern()) {
-            regexpScanner.scan(componentEffectiveKey, inputFile.file(), sourcesEncoding);
```
Comment (henryju) [inline sonar-scanner-engine/src/main/java/org/sonar/scanner/issue/ignore/scanner/IssueExclusionsLoader.java]: Nice bug fixed: we were not using the correct encoding!

File: sonar-scanner-engine/src/main/java/org/sonar/scanner/issue/ignore/scanner/IssueExclusionsRegexpScanner.java (hunk 3 above)
Comment (dbmeneses) [inline sonar-scanner-engine/src/main/java/org/sonar/scanner/issue/ignore/scanner/IssueExclusionsRegexpScanner.java]: ooops.. where is the sonar bot? :)

**Pull request merged**